File: frontend/src/types/api/bim.ts
export type BimFileDto = {
  id: string;
  dataset_id: string;
  filename: string;
  format: string;
  status: string | null;
  schema: string | null;
  stats: {
    storeys?: number;
    spaces?: number;
    entities?: number;
  } | null;
  units: Array<Record<string, unknown>> | null;
  created_at: string | null;
  extra: Record<string, unknown> | null;
};

export type BimMetadataDto = {
  id: string;
  filename: string;
  upload_date: string | null;
  size: number | null;
  content_type: string | null;
};

export type BimStoreyDto = {
  id: string;
  bim_dataset_id: string;
  global_id: string;
  name: string | null;
  elevation: number | null;
  created_at: string | null;
};

export type BimSpaceDto = {
  id: string;
  bim_dataset_id: string;
  global_id: string;
  name: string | null;
  raw_name: string | null;
  storey_id: string | null;
  area: number | null;
  volume: number | null;
  created_at: string | null;
};

File: backend/app/schemas/bim.py
from typing import Any

from pydantic import BaseModel, ConfigDict, Field


class BimFileResponse(BaseModel):
    model_config = ConfigDict(populate_by_name=True)

    id: str
    dataset_id: str
    filename: str
    format: str
    status: str | None
    bim_schema: str | None = Field(
        default=None,
        validation_alias="schema",
        serialization_alias="schema",
    )
    stats: dict[str, Any] | None
    units: list[dict[str, Any]] | None
    created_at: str | None
    extra: dict[str, Any] | None


class BimMetadataResponse(BaseModel):
    id: str
    filename: str
    upload_date: str | None
    size: int | None
    content_type: str | None


class BimStoreyResponse(BaseModel):
    id: str
    bim_dataset_id: str
    global_id: str
    name: str | None
    elevation: float | None
    created_at: str | None


class BimSpaceResponse(BaseModel):
    id: str
    bim_dataset_id: str
    global_id: str
    name: str | None
    raw_name: str | None
    storey_id: str | None
    area: float | None
    volume: float | None
    created_at: str | None
